### PR TITLE
[TC-287] Add support for per-DeliveryService routing names

### PR DESCRIPTION
**Pull request by rawlinp**
Add a routing_name column to the DeliveryService table and populate it
to 'tr' and 'edge' for HTTP and DNS delivery services, respectively.

To make upgrading seamless for CDNs using a custom http.routing.name
(configured in the TR), provide a temporary upgrade parameter to
populate the routing name for HTTP Delivery Services. Otherwise HTTP
DSes will get routing_name = "tr".

Replace all hardcoded/assumed routing names of 'edge' or 'tr' in Traffic
Ops with the proper routing name configured for that delivery service.

Update the API/UI to support custom routing names for delivery services.
To prevent breaking the API, "routingName" is an optional field (the DB
schema provides an arbitrary default).

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/138/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/99/

**Comment by dneuman64** [comment]
Thanks for squashing your commits, however a PR this size should probably have more than 1 commit.

**Comment by rawlinp** [comment]
@dneuman64 I apologize; this went through a handful of revisions which would've made the commit log really noisy. I'll try to split things up better next time which might make reviewing commit-by-commit easier for those who review that way.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-whatchanged/4/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-whatchanged/28/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/158/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/119/

File: traffic_ops/app/lib/Schema/Result/DeliveryServiceInfoForDomainList.pm [hunk 1]
```diff
@@ -86,6 +87,7 @@ __PACKAGE__->add_columns(
 	"multi_site_origin",           { data_type => "integer", is_nullable => 0 },
 	"ds_id",                       { data_type => "integer", is_nullable => 0 },
 	"dscp",                        { data_type => "integer", is_nullable => 0 },
+	"routing_name",                { data_type => "varchar", is_nullable => 0, size => 45 },
```
Comment (mitchell852) [inline traffic_ops/app/lib/Schema/Result/DeliveryServiceInfoForDomainList.pm]: is it 45 or 48? i saw this check in the api

+			routingName          => [ \&is_valid_routing_name, is_long_at_most( 48, 'too long' ) ],

File: traffic_ops/app/t/api/1.2/deliveryservice_crud.t [hunk 2]
```diff
@@ -124,6 +128,7 @@ sub run_ut {
 	        "regionalGeoBlocking" => "1",
 	        "active" => "false",
 	        "dscp" => 0,
+            "routingName" => "foo",
```
Comment (mitchell852) [inline traffic_ops/app/t/api/1.2/deliveryservice_crud.t]: can you check formatting on these. looks like you need one more tab.

File: traffic_portal/app/src/common/modules/form/deliveryService/form.deliveryService.Steering.tpl.html [hunk 3]
```diff
@@ -112,6 +112,16 @@
                 </div>
             </div>
 
+            <div class="form-group" ng-class="{'has-error': hasError(deliveryServiceForm.routingName), 'has-feedback': hasError(deliveryServiceForm.routingName)}">
```
Comment (mitchell852) [inline traffic_portal/app/src/common/modules/form/deliveryService/form.deliveryService.Steering.tpl.html]: these TP changes look good. is routingName not applicable to ANY_MAP delivery services?

File: docs/source/development/traffic_ops_api/v12/deliveryservice.rst [hunk 4]
```diff
@@ -171,6 +171,8 @@ Delivery Service
   +--------------------------+--------+--------------------------------------------------------------------------------------------------------------------------------------+
   | ``remapText``            | string | Additional raw remap line text.                                                                                                      |
   +--------------------------+--------+--------------------------------------------------------------------------------------------------------------------------------------+
+  | ``routingName``          | string | The routing name of this deliveryservice.                                                                                            |
```
[the remark this answers is not in the document]
Comment (rawlinp) [reply docs/source/development/traffic_ops_api/v12/deliveryservice.rst]: done, will update this in the next commit

File: docs/source/development/traffic_ops_api/v12/deliveryservice.rst [hunk 5]
```diff
@@ -1890,6 +1896,8 @@ URL Sig Keys
   +----------------------------------------+--------+----------+---------------------------------------------------------------------------------------------+
   | ``>routingType``                       | string | yes      | Eg. DNS or HTTP Redirect                                                                    |
   +----------------------------------------+--------+----------+---------------------------------------------------------------------------------------------+
+  | ``>routingName``                       | string | no       | The routing name for the delivery service, e.g. <routingName>.myds.cdn.com                  |
```
[the remark this answers is not in the document]
Comment (rawlinp) [reply docs/source/development/traffic_ops_api/v12/deliveryservice.rst]: done, will update this in the next commit

File: traffic_ops/app/lib/Schema/Result/DeliveryServiceInfoForDomainList.pm (hunk 1 above)
Comment (rawlinp) [reply traffic_ops/app/lib/Schema/Result/DeliveryServiceInfoForDomainList.pm]: Yeah, it's 48. I put 45 here to make it consistent with `xml_id`, which also has a check for 48 in the API but is 45 here. I'll set it to 48 for `routing_name` to match the API.

File: traffic_ops/app/t/api/1.2/deliveryservice_crud.t (hunk 2 above)
Comment (rawlinp) [reply traffic_ops/app/t/api/1.2/deliveryservice_crud.t]: yeah, the issue is there's a mix of tabs and spaces (on the same lines even) used in this file. I'll make it match.

File: traffic_portal/app/src/common/modules/form/deliveryService/form.deliveryService.Steering.tpl.html (hunk 3 above)
Comment (rawlinp) [reply traffic_portal/app/src/common/modules/form/deliveryService/form.deliveryService.Steering.tpl.html]: Thanks, I don't believe it is applicable. My understanding of ANY_MAP delivery services is that they're only meant to add raw lines to the remap config of edge caches they're assigned to.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/162/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/123/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-whatchanged/32/

File: traffic_portal/app/src/common/modules/form/deliveryService/form.deliveryService.Steering.tpl.html (hunk 3 above)
Comment (mitchell852) [reply traffic_portal/app/src/common/modules/form/deliveryService/form.deliveryService.Steering.tpl.html]: @elsloo - can you confirm that? that routingName is not applicable to ANY_MAP delivery services?

Comment (elsloo) [reply traffic_portal/app/src/common/modules/form/deliveryService/form.deliveryService.Steering.tpl.html]: @rawlinp is correct, ANY_MAPs do not need routing names, they aren't in the CRConfig and are intended to manually place remap rules on the edge.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/137/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-whatchanged/46/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/176/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-whatchanged/47/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/177/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/138/

**Comment by mitchell852** [comment]
@rawlinp and @dg4prez - i get this error when i run db/admin.pl upgrade:

$ db/admin.pl upgrade
Migrating database..
UPDATE deliveryservice ds
SET routing_name = (
  SELECT p.value
  FROM parameter p
    JOIN profile_parameter pp ON p.id = pp.parameter
    JOIN profile pro ON pp.profile = pro.id
    JOIN cdn ON pro.cdn = cdn.id
  WHERE p.name = 'upgrade_http_routing_name'
    AND cdn.id = ds.cdn_id)
WHERE routing_name IS NULL;
ERROR:  column "routing_name" does not exist
LINE 10: WHERE routing_name IS NULL

so i guess from that sql, i'm supposed to create one upgrade_http_routing_name parameter per profile per cdn or something?

**Comment by mitchell852** [comment]
not sure what happened there but db/admin.pl upgrade didn't add the routing_name column for me so i did it manually with 

ALTER TABLE deliveryservice ADD routing_name TEXT;

and all seems better now :)

**Comment by rawlinp** [comment]
@mitchell852 the `upgrade_http_routing_name` parameter is only necessary if you're using a custom `http.routing.name` for the traffic routers like 'ccr'. In that case you'd create that parameter and associate it to a profile for each CDN using a custom routing name. The value of the parameter should be the value of the custom routing name in use for that CDN.

**Pull request merged**